### PR TITLE
spdk: upgrade spdk to v16.12

### PR DESCRIPTION
**Pull request by liupan1111**
upgrade spdk to v16.12, upgrade dpdk to v16.11, and fix compilation issue.

### CODE REVIEW

**Comment by optimistyzy** [comment]
should be merged to support new spdk version in NVMEDecvice.cc

File: src/os/bluestore/NVMEDevice.cc [hunk 1]
```diff
@@ -223,6 +223,9 @@ class SharedDriverData {
   }
   ~SharedDriverData() {
     g_ceph_context->get_perfcounters_collection()->remove(logger);
+    if(!qpair) {
```
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: space after "if"

File: src/os/bluestore/NVMEDevice.cc [hunk 2]
```diff
@@ -587,41 +601,18 @@ static bool probe_cb(void *cb_ctx, struct spdk_pci_device *pci_dev)
...
-      derr << __func__ << " device " << name
-           << " " << spdk_pci_device_get_bus(pci_dev)
-           << ":" << spdk_pci_device_get_dev(pci_dev) << ":"
-           << spdk_pci_device_get_func(pci_dev)
-           << " bind to uio driver failed, may lack of uio_pci_generic kernel module" << dendl;
-      return false;
-    }
-  }
-
   return true;
```
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: how do we decide to use uio driver when removing above?

File: src/os/bluestore/NVMEDevice.cc [hunk 3]
```diff
@@ -554,31 +552,47 @@ class NVMEManager {
...
   NVMEManager::ProbeContext *ctx = static_cast<NVMEManager::ProbeContext*>(cb_ctx);
   char serial_number[128];
-  string name = spdk_pci_device_get_device_name(pci_dev) ? spdk_pci_device_get_device_name(pci_dev) : "Unknown";
-  dout(0) << __func__ << " found device at name: " << name
-          << " bus: " << spdk_pci_device_get_bus(pci_dev) << ":" << spdk_pci_device_get_dev(pci_dev) << ":"
+  struct spdk_pci_addr pci_addr;
+  struct spdk_pci_device *pci_dev = NULL;
+
+  if (trid->trtype != SPDK_NVME_TRANSPORT_PCIE) {
+    // currently, only probe local nvme device.
```
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: plz log enough information

File: src/os/bluestore/NVMEDevice.cc [hunk 4]
```diff
@@ -554,31 +552,47 @@ class NVMEManager {
...
+    return false;
+  }
+
+  if (spdk_pci_addr_parse(&pci_addr, trid->traddr)) {
+    return false;
```
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: plz log enough information

File: src/os/bluestore/NVMEDevice.cc [hunk 5]
```diff
@@ -554,31 +552,47 @@ class NVMEManager {
...
+  }
+
+  pci_dev = spdk_pci_get_device(&pci_addr);
+  if (!pci_dev) {
```
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: plz log enough information

File: src/os/bluestore/NVMEDevice.cc [hunk 6]
```diff
@@ -201,10 +198,13 @@ class SharedDriverData {
         flush_lock("NVMEDevice::flush_lock"),
         flush_waiters(0),
         completed_op_seq(0), queue_op_seq(0) {
+    enum spdk_nvme_qprio qprio = SPDK_NVME_QPRIO_URGENT;
```
Comment (yuyuyu101) [inline src/os/bluestore/NVMEDevice.cc]: @optimistyzy it this "URGENT" properly used here?

**Comment by yuyuyu101** [comment]
good move

File: src/os/bluestore/NVMEDevice.cc (hunk 1 above)
Comment (liupan1111) [inline src/os/bluestore/NVMEDevice.cc]: agree

File: src/os/bluestore/NVMEDevice.cc (hunk 5 above)
Comment (liupan1111) [inline src/os/bluestore/NVMEDevice.cc]: sure，thanks.

File: src/os/bluestore/NVMEDevice.cc (hunk 2 above)
Comment (optimistyzy) [inline src/os/bluestore/NVMEDevice.cc]: The logic is in SPDK.  We will either use VFIO or UIO, the logic is shown in  scripts/setup.sh.  For pci_access,  the support is removed in current SPDK.

File: src/os/bluestore/NVMEDevice.cc (hunk 6 above)
Comment (optimistyzy) [inline src/os/bluestore/NVMEDevice.cc]: Yes,  the default is URGENT(0). There are three values defined in SPDK, and the default priority is 0. It is correctly used in this scenario

**Comment by liupan1111** [comment]
@yuyuyu101 modified, please take a look, thanks!

**Pull request merged**